File: module-10-fine-tuning/solutions/train.ts
/**
 * Module 10: Training orchestrator
 * Orchestrates fine-tuning jobs (SFT, DPO).
 *
 * This module demonstrates the pipeline. Actual training requires
 * a Fireworks/OpenAI-compatible API endpoint.
 */
import "dotenv/config";
import * as fs from "fs/promises";
import * as path from "path";
import { extractTraces, clusterTraces, stratifiedSample, saveTraces } from "./data-pipeline.js";
import { buildSFTDataset, saveSFTDataset } from "./sft-dataset.js";
import { buildDPODataset, saveDPODataset } from "./dpo-dataset.js";
import { routeIntent } from "./classifier.js";
import { CODING_AGENT_PROMPT } from "../../module-5-coding-agent/solutions/prompt.js";
import type { Trial } from "../../module-7-evals/solutions/types.js";

const OUTPUT_DIR = path.join(process.cwd(), "training-data");

async function main() {
  await fs.mkdir(OUTPUT_DIR, { recursive: true });

  // In a real pipeline, load trials from eval results
  // For demonstration, we create synthetic trial data
  console.log("=== Module 10: Fine-Tuning Pipeline ===\n");

  console.log("Step 1: Loading eval traces...");
  // Load from eval report if available
  let trials: Trial[] = [];
  try {
    const reportPath = path.join(process.cwd(), "eval-report-cycle-1.json");
    const reportData = await fs.readFile(reportPath, "utf-8");
    const report = JSON.parse(reportData);
    trials = report.tasks.flatMap((t: { trials: Trial[] }) => t.trials);
    console.log(`  Loaded ${trials.length} trials from eval report`);
  } catch {
    console.log("  No eval report found. Run module-7 evals first to generate training data.");
    console.log("  Demonstrating pipeline with synthetic data...\n");

    // Create minimal synthetic data for demonstration
    trials = createSyntheticTrials();
  }

  // Step 2: Extract and cluster traces
  console.log("\nStep 2: Extracting and clustering traces...");
  const traces = extractTraces(trials);
  await saveTraces(traces, path.join(OUTPUT_DIR, "traces.jsonl"));

  const clusters = clusterTraces(traces);
  console.log(`  Found ${clusters.size} clusters`);

  const sampled = stratifiedSample(clusters);
  console.log(`  Sampled ${sampled.length} representative traces`);

  // Step 3: Build SFT dataset
  console.log("\nStep 3: Building SFT dataset...");
  const sftExamples = buildSFTDataset(traces, CODING_AGENT_PROMPT);
  await saveSFTDataset(sftExamples, path.join(OUTPUT_DIR, "sft-dataset.jsonl"));

  // Step 4: Build DPO dataset
  console.log("\nStep 4: Building DPO dataset...");
  const dpoPairs = buildDPODataset(traces);
  await saveDPODataset(dpoPairs, path.join(OUTPUT_DIR, "dpo-dataset.jsonl"));

  // Step 5: Demonstrate classifier
  console.log("\nStep 5: Intent classification demo...");
  const testTasks = [
    "Fix the null pointer bug in auth.ts",
    "Create a new config file for the database",
    "What does the login function do?",
  ];
  for (const task of testTasks) {
    const result = await routeIntent(task);
    console.log(`  "${task.slice(0, 50)}" → ${result.label} (${(result.confidence * 100).toFixed(0)}%)`);
  }

  console.log(`\n✓ Training data saved to ${OUTPUT_DIR}/`);
  console.log("  - traces.jsonl: Raw trace records");
  console.log("  - sft-dataset.jsonl: SFT training examples");
  console.log("  - dpo-dataset.jsonl: DPO preference pairs");
  console.log("\nTo fine-tune, upload these datasets to Fireworks AI or OpenAI.");
}

function createSyntheticTrials(): Trial[] {
  // Minimal synthetic data for pipeline demonstration
  return [
    {
      taskId: "rename-variable",
      trialIndex: 0,
      transcript: {
        task: "Rename variable x to count",
        turns: [
          { role: "user", content: "Rename variable x to count" },
          {
            role: "assistant",
            content: "I'll rename the variable.",
            toolCalls: [
              { name: "read-file", input: { path: "app.ts" }, result: "const x = 0;" },
              { name: "edit-file", input: { path: "app.ts", old_string: "x", new_string: "count" }, result: "Edited" },
            ],
          },
        ],
        totalTokens: 500,
        durationMs: 3000,
      },
      grade: { score: 1, passed: true, explanation: "Variable renamed correctly" },
    },
    {
      taskId: "rename-variable",
      trialIndex: 1,
      transcript: {
        task: "Rename variable x to count",
        turns: [
          { role: "user", content: "Rename variable x to count" },
          { role: "assistant", content: "Done.", toolCalls: [] },
        ],
        totalTokens: 200,
        durationMs: 1000,
      },
      grade: { score: 0, passed: false, explanation: "Did not actually edit the file" },
    },
  ];
}

main().catch(console.error);

File: module-10-fine-tuning/solutions/python/train.py
"""Module 10: Training orchestrator."""
import os
import sys
import json

sys.path.insert(0, os.path.dirname(__file__))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-5-coding-agent", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-7-evals", "solutions", "python"))

from dotenv import load_dotenv
load_dotenv()

from data_pipeline import extract_traces, cluster_traces, stratified_sample, save_traces
from sft_dataset import build_sft_dataset, save_sft_dataset
from dpo_dataset import build_dpo_dataset, save_dpo_dataset
from classifier import route_intent
from prompt import CODING_AGENT_PROMPT
from eval_types import Trial, Transcript, TranscriptTurn, GradeResult

OUTPUT_DIR = os.path.join(os.getcwd(), "training-data")


def create_synthetic_trials() -> list[Trial]:
    return [
        Trial(
            task_id="rename-variable",
            trial_index=0,
            transcript=Transcript(
                task="Rename variable x to count",
                turns=[
                    TranscriptTurn(role="user", content="Rename variable x to count"),
                    TranscriptTurn(role="assistant", content="I'll rename the variable.", tool_calls=[
                        {"name": "read-file", "input": {"path": "app.ts"}, "result": "const x = 0;"},
                        {"name": "edit-file", "input": {"path": "app.ts", "old_string": "x", "new_string": "count"}, "result": "Edited"},
                    ]),
                ],
                total_tokens=500, duration_ms=3000,
            ),
            grade=GradeResult(score=1.0, passed=True, explanation="Variable renamed correctly"),
        ),
        Trial(
            task_id="rename-variable",
            trial_index=1,
            transcript=Transcript(
                task="Rename variable x to count",
                turns=[
                    TranscriptTurn(role="user", content="Rename variable x to count"),
                    TranscriptTurn(role="assistant", content="Done."),
                ],
                total_tokens=200, duration_ms=1000,
            ),
            grade=GradeResult(score=0.0, passed=False, explanation="Did not actually edit the file"),
        ),
    ]


def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    print("=== Module 10: Fine-Tuning Pipeline ===\n")

    print("Step 1: Loading eval traces...")
    trials = []
    try:
        report_path = os.path.join(os.getcwd(), "eval-report-cycle-1.json")
        with open(report_path) as f:
            report = json.load(f)
        # In real pipeline, would reconstruct Trial objects from report
        print(f"  Found eval report, but using synthetic data for demo")
        trials = create_synthetic_trials()
    except FileNotFoundError:
        print("  No eval report found. Using synthetic data for demo...")
        trials = create_synthetic_trials()

    print("\nStep 2: Extracting and clustering traces...")
    traces = extract_traces(trials)
    save_traces(traces, os.path.join(OUTPUT_DIR, "traces.jsonl"))

    clusters = cluster_traces(traces)
    print(f"  Found {len(clusters)} clusters")

    sampled = stratified_sample(clusters)
    print(f"  Sampled {len(sampled)} representative traces")

    print("\nStep 3: Building SFT dataset...")
    sft_examples = build_sft_dataset(traces, CODING_AGENT_PROMPT)
    save_sft_dataset(sft_examples, os.path.join(OUTPUT_DIR, "sft-dataset.jsonl"))

    print("\nStep 4: Building DPO dataset...")
    dpo_pairs = build_dpo_dataset(traces)
    save_dpo_dataset(dpo_pairs, os.path.join(OUTPUT_DIR, "dpo-dataset.jsonl"))

    print("\nStep 5: Intent classification demo...")
    test_tasks = [
        "Fix the null pointer bug in auth.ts",
        "Create a new config file for the database",
        "What does the login function do?",
    ]
    for task in test_tasks:
        result = route_intent(task)
        print(f'  "{task[:50]}" -> {result["label"]} ({result["confidence"] * 100:.0f}%)')

    print(f"\nTraining data saved to {OUTPUT_DIR}/")
    print("  - traces.jsonl: Raw trace records")
    print("  - sft-dataset.jsonl: SFT training examples")
    print("  - dpo-dataset.jsonl: DPO preference pairs")
    print("\nTo fine-tune, upload these datasets to Fireworks AI or OpenAI.")


if __name__ == "__main__":
    main()
